Add a small explanation on why Restore is needed. (#563)
open-telemetry/opentelemetry-specification: specification/context/context.md

The API MUST return a value that can be used as a `Token` to restore the previous
`Context`.

> Note that every call to this operation should result in a corresponding call to

Reviewer: Strongly disagree with this. Normally I would assume that the next context is just attached to the execution unit, overwriting the previous. The context will not have a back-reference to the execution unit anyway.

Author: It might be complicated but that's going to help users prevent errors on their side. Observe this is already done in `io.grpc.Context` and also in `contextvars` (through https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.reset)

So I think users don't *have* to actually call `Detach` for every `Attach`, but ideally they *should*.

Reviewer: How is Python's reset related to this Detach?

Author: It's semantically similar, i.e. `Reset the context variable to the value it had before the ContextVar.set() that created the token was used.`

Although, of course, it's not bringing back the entire `Context` :/ Wording it as "Note that it is recommended to have a corresponding call to Detach Context" would help? (by making it less mandatory)

Author: Ping @Oberon00  ;)

Reviewer: Python's token.reset is quite trivial, as the token stores the old value and `var.reset(token)` is basically (minus error checking & single-use) just an alias for var.set(token.old_value) (see [source](https://github.com/python/cpython/blob/v3.8.2/Python/context.c#L304-L309)). I'm not entirely sure about detach. Is it to be called as `ContextModule.Detach(prevcontext)`? If so it can be similarly easy because it could similarly be an alias for `ContextModule._internal_set_context(prevcontext.parent)` while `ContextModule.Attach(ctx)` would do `ctx.parent = ContextModule.Current(); ContextModule._internal_set_context(ctx)`.

Author: @Oberon00 The idea is that, yes, `Detach Context` will have a very similar semantic to what you described (equivalent to `set(prevContext)` plus the additional checks).

So for the `Context` routines we have tried to be more 'flexible' as quite often we will be using language-specific implementations, and we can't change them. For `Detach` it can be implemented in any equivalent way. For example, in Java we have:

```java
// io.grpc.Context
currentContext.detach(prevContext);

// Or, use the sugar we offer in OTel
try (Scope scope =  ContextUtils.withScopedContext(context)) {
  // Context.detach(prevContext) will be automatically called.
}
```

Reviewer 2: @Oberon00 happy with this?

[Detach Context](#detach-context).

### Detach Context

Resets the `Context` associated with the caller's current execution unit
to the value it had before attaching a specified `Context`.

This operation is intended to help making sure the correct `Context`
is associated with the caller's current execution unit. Users can
rely on it to identify a wrong call order and emit a signal to warn

> users about the broken invariant.

Reviewer 3: > Users can rely on it to identify a wrong call order and emit a signal to warn users about the broken invariant.

Is a bit unclear to me. I'm reading between the lines, but guessing:

Wrong call order means:

because there is a pattern that an attach of a context should be paired with a detach, then someone reading the code will be able to spot the lifetime of the context. In this regard, someone can quickly identify if the context is detached in the wrong location

"emit a signal to warn users about the broken invariant" I'm having a hard time guessing. What signal is emitted? I'm assuming "broken invariant" is referring to the context object.

Author: @toumorokoshi Updated it to make it a little bit clearer: so basically

1) The wrong call order: trying the incorrect Context (usually not the current instance)
2) Emit a signal: either logging an error or return a value (this is implementation specific).

Let me know ;)

Author: Ping @toumorokoshi

Author: Ping @toumorokoshi Let me me know if this is better now ;)

Reviewer 3: I think the examples actually help the most. But either way good on merging it and not waiting for my very slow feedback  :)

The API MUST accept the following parameters:

> - A `Token` that was returned by a previous call to attach a `Context`.

Reviewer 4: Does the operation it return nothing or should it return information on whether the operation was successful or failed because the respective context was already detached again or is not on top of stack?

Author: I think it returns nothing in all the languages/implementations I've seen. We could probably say something along 'It CAN optionally return additional information on the success or failure, etc".

The API MUST return a value that can be used as a `Token` to restore the previous
`Context`.

Note that every call to this operation should result in a corresponding call to
[Detach Context](#detach-context).

### Detach Context

Resets the `Context` associated with the caller's current execution unit
to the value it had before attaching a specified `Context`.

This operation is intended to help making sure the correct `Context`
is associated with the caller's current execution unit. Users can
rely on it to identify a wrong call order, i.e. trying to detach
a `Context` that is not the current instance. In this case the operation
can emit a signal to warn users of the wrong call order, such as logging
an error or returning an error value.

The API MUST accept the following parameters:

- A `Token` that was returned by a previous call to attach a `Context`.

> The API optionally CAN return a value used to check whether the operation

Reviewer 4: ```suggestion
The API MAY return a value used to check whether the operation
```
